liquidity_pool/src/lib.rs

```diff
@@ -1,487 +1,341 @@
+            "position can't be liquidated, not enough tokens sent"
+        );
+
+        // Refund extra tokens paid
+        let extra_payment = &payment_amount - &total_owed;
+        if extra_payment > 0 {
+            match self.send().direct_esdt_via_transf_exec(
+                &caller,
+                token_id.as_esdt_identifier(),
+                &extra_payment,
+                &[],
+            ) {
+                Result::Ok(()) => {}
+                Result::Err(_) => return sc_error!("Failed refunding extra tokens"),
+            };
+        }
 
-        self.debt_positions()
-            .insert(position_id, debt_position.clone());
+        self.debt_position(position_id)
+            .update(|d| d.is_liquidated = true);
 
-        let liquidate_data = LiquidateData {
-            collateral_token: debt_position.collateral_identifier,
-            amount,
-        };
+        // decrease circulating supply
+        self.total_circulating_supply()
+            .update(|circulating_supply| *circulating_supply -= &payment_amount);
 
-        Ok(liquidate_data)
+        Ok(())
     }
 
     #[payable("EGLD")]
-    #[endpoint]
-    fn issue(
+    #[endpoint(issueStablecoinToken)]
+    fn issue_stablecoin_token(
         &self,
-        plain_ticker: BoxedBytes,
-        token_ticker: TokenIdentifier,
-        token_prefix: BoxedBytes,
         #[payment] issue_cost: BigUint,
     ) -> SCResult<AsyncCall<BigUint>> {
         only_owner!(self, "only owner can issue new tokens");
         require!(
-            issue_cost == BigUint::from(ESDT_ISSUE_COST),
-            "payment should be exactly 5 EGLD"
-        );
-        require!(
-            token_ticker == self.pool_asset().get(),
-            "wrong ESDT asset identifier"
+            self.stablecoin_token_id().is_empty(),
+            "Stablecoin already issued"
         );
 
-        let issue_data = self.prepare_issue_data(token_prefix.clone(), plain_ticker);
-        require!(
-            issue_data.name != BoxedBytes::zeros(0),
-            "invalid input. could not prepare issue data"
-        );
-        require!(
-            issue_data.is_empty_ticker,
-            "token already issued for this identifier"
-        );
+        let token_display_name = BoxedBytes::from(STABLE_COIN_NAME);
+        let token_ticker = BoxedBytes::from(STABLE_COIN_TICKER);
+        let initial_supply = BigUint::from(1u32);
 
         Ok(ESDTSystemSmartContractProxy::new()
-            .issue_semi_fungible(
+            .issue_fungible(
                 issue_cost,
-                &issue_data.name,
-                &BoxedBytes::from(issue_data.ticker.as_esdt_identifier()),
-                SemiFungibleTokenProperties {
+                &token_display_name,
+                &token_ticker,
+                &initial_supply,
+                FungibleTokenProperties {
+                    can_burn: true,
+                    can_mint: true,
+                    num_decimals: 0,
```

Comment:
why 0 decimals ?

Reply:
Will probably have to align with BASE_PRECISION, i.e. 9, but will see about that. Good catch.